### PR TITLE
[processing] Allow algorithms to specify tags

### PR DESCRIPTION
**Pull request by nyalldawson**
Tags are used while searching in the toolbox. This should help with finding algorithms when the exact name is not known, eg you could search for "envelope" or "bounds" and find the 'Polygon from Layer Extent' algorithm.

At the moment it's quite hard to discover algorithms which exist when you don't know what their called and have to instead search for every possible naming variant which could exist...

### CODE REVIEW

**Comment by nirvn** [comment]
Nice, should be really useful. 

It will however further increase the chances of someone stumbling on "duplicate" algs from different providers, which is confusing for newcomers and makes it more difficult for fellow QGIS users to offer remote assistance.

@volaya , what happened to the normal vs. advanced view for toolbox? IMHO it would be good to refine that idea. For eg in the normal mode, duplicate algs from different providers should be hidden, favoring a/ QGIS algs, and b/ most stable options (ie QGIS' clip function is much less prone to fail than SAGA).

**Comment by nyalldawson** [comment]
> It will however further increase the chances of someone stumbling on "duplicate" algs from different providers, which is confusing for newcomers and makes it more difficult for fellow QGIS users to offer remote assistance.

Agreed, but it's a separate issue and shouldn't affect the mergability of this PR.

**Comment by nirvn** [comment]
@nyalldawson yep, you're right. I was just being opportunistic :-P

**Comment by pcav** [comment]
Very useful, thanks. Tags had a very limited success for plugins, as there are a number of very similar or too generic tags (vector/vectors, etc.). May I suggest to present a number of standard tags, so devs may converge on those unless they have special needs?

**Comment by nyalldawson** [comment]
@volaya thanks! I'll start adding tags I think are sensible to existing algs

**Comment by volaya** [comment]
@nyalldawson great!

File: python/plugins/processing/modeler/ModelerDialog.py [hunk 1]
```diff
@@ -486,7 +487,15 @@ def fillAlgorithmTreeUsingProviders(self):
                     continue
                 if alg.commandLineName() == self.alg.commandLineName():
                     continue
-                if text == '' or text.lower() in alg.name.lower():
+
+                item_text = [alg.name.lower()]
+                item_text.extend(alg.tags.split(','))
+
+                show = not search_strings or all(
```
Comment (luipir) [inline python/plugins/processing/modeler/ModelerDialog.py]: clear the result, hard (for me) to read the code. I'm not fan of pythonic syntax, and I would prefer explicit and readable code if no performance problems are involved. Is it possible to reformulate?

Comment (nyalldawson) [inline python/plugins/processing/modeler/ModelerDialog.py]: I agree... but then.... I'm struggling to come up with an alternative version which does the same without just expanding all the loops out

**Comment by NathanW2** [comment]
I think it's fine.

**Pull request merged**